### DISCUSSION TITLE
How to get the most out of vim?

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [zopencommunity/meta#65] [Reference]
I have to admit I've not used vim, but I see it apparently is now available with unknown quality: https://github.com/ZOSOpenTools/vimport/releases/tag/vimport_226

Any people that use 'vim' out there that would like to provide some input on perhaps a good config to start with?

**Comment by jshimoda9323** [comment]
I have shamelessly copied and have used these settings for a while from somewhere....I cannot remember the original source, but it seems to work fine for me.  You may want to play with the tabstop, expandtab, shiftwidth, and autoindent settings to suit your indentation needs.  Put this in the `.vimrc` file in your home directory.  Also, the color eldar setting may need additional files?  Just comment that out if it doesn't work.
```
" Use Vim settings, rather then Vi settings (much better!).
" This must be first, because it changes other options as a side effect.
set nocompatible

" allow backspacing over everything in insert mode
set backspace=indent,eol,start

set history=50          " keep 50 lines of command line history
set ruler               " show the cursor position all the time
set showcmd             " display incomplete commands
set incsearch           " do incremental searching
set expandtab
set tabstop=2
set shiftwidth=2
set nobackup
set autoindent
"set cindent
set ic

color eldar

" Don't use Ex mode, use Q for formatting
map Q gq

" CTRL-U in insert mode deletes a lot.  Use CTRL-G u to first break undo,
" so that you can undo CTRL-U after inserting a line break.
inoremap <C-U> <C-G>u<C-U>

" In many terminal emulators the mouse works just fine, thus enable it.
if has('mouse')
  set mouse=a
endif

" Switch syntax highlighting on, when the terminal has colors
" Also switch on highlighting the last used search pattern.
if &t_Co > 2 || has("gui_running")
  syntax on
  set hlsearch
endif

" Only do this part when compiled with support for autocommands.
if has("autocmd")

  " Enable file type detection.
  " Use the default filetype settings, so that mail gets 'tw' set to 72,
  " 'cindent' is on in C files, etc.
  " Also load indent files, to automatically do language-dependent indenting.
  "filetype plugin indent on

  " Put these in an autocmd group, so that we can delete them easily.
  augroup vimrcEx
  au!

  " For all text files set 'textwidth' to 78 characters.
  autocmd FileType text setlocal textwidth=78

  " When editing a file, always jump to the last known cursor position.
  " Don't do it when the position is invalid or when inside an event handler
  " (happens when dropping a file on gvim).
  " Also don't do it when the mark is in the first line, that is the default
  " position when opening a file.
  autocmd BufReadPost *
    \ if line("'\"") > 1 && line("'\"") <= line("$") |
    \   exe "normal! g`\"" |
    \ endif

  augroup END

else

  set autoindent                " always set autoindenting on

endif " has("autocmd")

" Convenient command to see the difference between the current buffer and the
" file it was loaded from, thus the changes you made.
" Only define it when not defined already.
if !exists(":DiffOrig")
  command DiffOrig vert new | set bt=nofile | r # | 0d_ | diffthis
                  \ | wincmd p | diffthis
endif

set visualbell
```

**Reply by MikeFultonDev** [reply]
worked for me once i deleted the color eldar line...
now i want to know what that gives me :)

how do i get colorization for C code?

**Reply by IgorTodorovskiIBM** [reply]
They should be in the VIM runtime under vim/share/vim/vim90/syntax/cpp.vim

I think there's a mistake in setting $VIM. It should point to $PWD/share/vim/vim90/, but then that reveals a slew of errors when I tried it.

**Reply by IgorTodorovskiIBM** [reply]
Ah, I had vim aliased to vi.  Setting VIM correctly should fix it.

**Reply by IgorTodorovskiIBM** [reply]
If you type:
`:color <tab>` during your vim session you can test out the various colours

**Reply by MikeFultonDev** [reply]
I noticed that the code in @jshimoda9323 's example:
```
if &t_Co > 2 || has("gui_running")
  syntax on
  set hlsearch
endif
```

isn't activated, but if I turn it on, I get underling, bold etc. in the 'right' places for C. 
If I want colourful code, should I be putting a `set color` command of some form in?

**Reply by jshimoda9323** [reply]
The `syntax on` setting is the thing that gives you syntax coloring.  The `color <color_name>` setting just sets a 'template' of colors that the syntax highlighter will choose from.  The `color eldar` in my vim settings is a color template I found once that looked good to me :)

**Comment by MikeFultonDev** [comment]
Lots of good info plus an intro here: https://zosopentools.github.io/meta/#/Guides/VimOnZOS so moving this to reference.

**Comment by MikeFultonDev** [comment]
Closing this since the doc section for vim was created
